keps/sig-node/3857-rro-hostpath/README.md

```diff
@@ -0,0 +1,1046 @@
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with `recursivelyReadOnly: Enabled` will not work on old distros (See "Constraints" above).
+
+Mitigation: set `recursivelyReadOnly: BestEffort`.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Core API
+Add `RecursivelyReadOnly: (Disabled|BestEffort|Enabled)` to the [`VolumeMount`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/pkg/apis/core/types.go#L1854-L1880) struct:
+- `Disabled` (default): always non-recursive
+- `BestEffort`: recursive if possible, otherwise non-recursive, without raising an error.
+   This value is translated to `Enabled` or `Disabled` in the `VolumeMount` struct (see the "Diff" below).
+- `Enabled`: recursive if possible, otherwise raise an error (if the node components are aware of this feature).
+
+When `RecursivelyReadOnly` is set to any non-nil value, `ReadOnly` is automatically implied to be `true`.
+When `RecursivelyReadOnly` is set to `BestEffort` or `Enabled`, `MountPropagation` should be set to `"None"` (default).
+
+A pod manifest will look like this:
+```yaml
+spec:
+  volumes:
+    - name: foo
+      hostPath: 
+        path: /mnt
+        type: Directory
+  containers:
+  - volumeMounts:
+    - mountPath: /mnt
+      name: foo
+      mountPropagation: None
+      # NEW
+      recursivelyReadOnly: BestEffort
+```
+
+`RecursivelyReadOnly` is (probably) specific to hostPath mounts, so it could be rather added to [`HostPathVolumeSource`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/pkg/apis/core/types.go#L662-L666) struct.
+However, as the `VolumeMount` struct already has `MountPropagation`, `RecursivelyReadOnly` should be added to the `VolumeMount` struct too to keep the design consistent.
+
+Diff:
+```diff
+diff --git a/pkg/apis/core/types.go b/pkg/apis/core/types.go
+index 04ffd250b61..21ca8e3e006 100644
+--- a/pkg/apis/core/types.go
++++ b/pkg/apis/core/types.go
+@@ -1857,7 +1857,7 @@ type VolumeMount struct {
+        Name string
+        // Optional: Defaults to false (read-write).
+        // +optional
+-       ReadOnly bool
++       ReadOnly *bool
+        // Required. If the path is not an absolute path (e.g. some/path) it
+        // will be prepended with the appropriate root prefix for the operating
+        // system.  On Linux this is '/', on Windows this is 'C:\'.
+@@ -1878,6 +1878,9 @@ type VolumeMount struct {
+        // SubPathExpr and SubPath are mutually exclusive.
+        // +optional
+        SubPathExpr string
++       // When RecursivelyReadOnly is set to a non-nil value, ReadOnly is implied to be true.
```

Comment:
Need to spec the default value (If not specifed, this defaults to Disabled).

Reply:
Done